src/algebra/order/upper_lower.lean

```diff
@@ -0,0 +1,192 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import algebra.order.group.basic
+import order.upper_lower
+
+/-!
+# Algebraic operations on upper/lower sets
+
+Upper/lower sets are preserved under pointwise algebraic operations in ordered groups.
+-/
+
+open function set
+open_locale pointwise
+
+section ordered_comm_monoid
+variables {α : Type*} [ordered_comm_monoid α] {s : set α} {x : α}
+
+@[to_additive] lemma is_upper_set.smul_subset (hs : is_upper_set s) (hx : 1 ≤ x) : x • s ⊆ s :=
+smul_set_subset_iff.2 $ λ y, hs $ le_mul_of_one_le_left' hx
+
+@[to_additive] lemma is_lower_set.smul_subset (hs : is_lower_set s) (hx : x ≤ 1) : x • s ⊆ s :=
+smul_set_subset_iff.2 $ λ y, hs $ mul_le_of_le_one_left' hx
+
+end ordered_comm_monoid
+
+section ordered_comm_group
+variables {α : Type*} [ordered_comm_group α] {s t : set α} {a : α}
+
+@[to_additive] lemma is_upper_set.smul (hs : is_upper_set s) : is_upper_set (a • s) :=
+begin
+  rintro _ y hxy ⟨x, hx, rfl⟩,
+  exact mem_smul_set_iff_inv_smul_mem.2 (hs (le_inv_mul_iff_mul_le.2 hxy) hx),
+end
+
+@[to_additive] lemma is_lower_set.smul (hs : is_lower_set s) : is_lower_set (a • s) :=
+hs.of_dual.smul
+
+@[to_additive] lemma set.ord_connected.smul (hs : s.ord_connected) : (a • s).ord_connected :=
+begin
+  rw [←hs.upper_closure_inter_lower_closure, smul_set_inter],
+  exact (upper_closure _).upper.smul.ord_connected.inter (lower_closure _).lower.smul.ord_connected,
+end
+
+@[to_additive] lemma is_upper_set.mul_left (ht : is_upper_set t) : is_upper_set (s * t) :=
+by { rw [←smul_eq_mul, ←bUnion_smul_set], exact is_upper_set_Union₂ (λ x hx, ht.smul) }
+
+@[to_additive] lemma is_upper_set.mul_right (hs : is_upper_set s) : is_upper_set (s * t) :=
+by { rw mul_comm, exact hs.mul_left }
+
+@[to_additive] lemma is_lower_set.mul_left (ht : is_lower_set t) : is_lower_set (s * t) :=
+ht.of_dual.mul_left
+
+@[to_additive] lemma is_lower_set.mul_right (hs : is_lower_set s) : is_lower_set (s * t) :=
+hs.of_dual.mul_right
+
+@[to_additive] lemma is_upper_set.div_left (ht : is_upper_set t) : is_lower_set (s / t) :=
```

Comment:
Could we also have an `inv` version? Maybe with `set.image`/`set.preimage` inv instead